Implement edge centrality calculations

koebi opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box when you have checked you have done them): -->
- [x] 1. I have [**rebased**](https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines) the latest version of the master branch into my feature branch and all conflicts have been resolved.
- [ ] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [x] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage and why the change was needed.

### Information about the changes
The centrality endpoint now supports two modes: "edge" and "node" mode, returning edge and node centrality respectively.

koebi committed:
restructure test data setup, add test for wrong size coordinates and bbox

koebi committed:
remove whitespace, fix typo

koebi committed:
run centrality parameter tests on smaller bounding box to stay within usual API test time frame

koebi committed:
implement edge-centrality calculation up to creating a valid, usable JSONresponse

koebi committed:
include changes in algorithm

koebi committed:
implement centrality response according to design document

koebi committed:
add param tests for edge-centrality endpoint

koebi committed:
only output 2D coordinates rounded to 6 decimal places

koebi committed:
adapt node centrality tests to new algorithm routines

koebi committed:
Correctly initialize EdgeExplorer

As stated in the graphhopper documentation, only one instance of EdgeExplorer
should ever be created. This instance is then transferred to the
CentralityAlgorithm, since it is a central part of its workings.

The EdgeExplorer used is set up to only accept outgoing edges valid for
the flagEncoder derived from the given profile.

koebi committed:
Initialize edge centrality values correctly

As the original algorithm does, all present edges get initialized such
that zero-values are reported correctly.

koebi committed:
Explicitly create directed graph

As noted by graphhopper, edge(a, b, dist, bothDirections) is
deprecated and edge().setDistance().setFlags() is used instead.

As the used flagEncoder does not store different speeds for reverse
edges, instead of using backwards edges, use forward edges with reversed
base nodes.

koebi committed:
Add tests for edge centrality

expectedScores were calculated via networkx

takb committed:
Merge branch 'master' into centrality

takb committed:
Merge branch 'master' into centrality

koebi committed:
Log only if validationFails()

koebi committed:
Merge branch 'centrality' of github.com:GIScience/openrouteservice into centrality

takb committed:
Merge branch 'master' into centrality

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -17,30 +17,59 @@
+        upperRight.put(49.421080);
+        tooLargeCoordinate.put(8.677139);
+        tooLargeCoordinate.put(49.395446);
+        tooLargeCoordinate.put(49.395446);
```
Should the third one be `tooLarge...` or `tooSmall...`? When reading, it is unclear, whether too large (or small) means 'too large (or small) values' in the sense that the coordinates are outside of the defined area, or it means that coordinates conain too many (or few) elements. Or even something else? Please clarify this in the code.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -54,7 +83,7 @@ public ParamsTest() {
     @Test
     public void testBasicFunctionality() {
```
What is "basic functionality"? The name of a test should describe the requirement it is testing. The comment in line 82 can be removed, as it explains nothing.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -66,16 +95,16 @@ public void testBasicFunctionality() {
 
     // test that excludeNodes get excluded
     @Test
     public void testExcludeNodes() {
```
Again the name of the test does not explain the tested requirement. You might inspire yourself from the comment above, which can then be removed.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -66,16 +95,16 @@ public void testBasicFunctionality() {
-        body.put("bbox", getParameter("heidelberg"));
+        body.put("bbox", getParameter("neuenheimBox"));
 
         // check that nodes to exclude are present in the response
```
Unclear comment: if the nodes should be excluded, why should they be present in the response?

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -84,17 +113,17 @@ public void testExcludeNodes() {
         int node1 = nodeIds.get(1);
         int node2 = nodeIds.get(2);
         int node3 = nodeIds.get(3);
+        System.out.printf("%d, %d, %d, %d\n", node0, node1, node2, node3);
```
In general we do not want to system.out anything. Rather use a logger or remove the line.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -120,7 +149,7 @@ public void testExcludeNodes() {
     @Test
     public void testInvalidExcludeNodes() {
```
Again, please find a better name for the test.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -133,17 +162,64 @@ public void testInvalidExcludeNodes() {
-    // + too few coords
+    // test that invalid bounding boxes with too few/too many coordinates return an error
+    @Test
+    public void testWrongSizeBBox() {
```
Again, prefer a good name over a comment.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -133,17 +162,64 @@ public void testInvalidExcludeNodes() {
+    // test that invalid bounding boxes with too large/too small coordinates return an error
     @Test
-    public void testInvalidBBox() {
+    public void testWrongSizedCoordinates() {
```
Naming again: what is this this test supposed to test?

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -157,8 +233,94 @@ public void testInvalidBBox() {
+    // test that empty bounding boxes get handled appropriately
+    //TODO: currently, both locations and nodeScores will just return empty.
+    // This is the expected and valid answer, but not too helpful.
+    // Should some kind of error be reported to the user?
```
This question should be clarified in the team meeting.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -157,8 +233,94 @@ public void testInvalidBBox() {
+    // This is the expected and valid answer, but not too helpful.
+    // Should some kind of error be reported to the user?
+    @Test
+    public void testEmptyBbox() {
```
Again naming.

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -157,8 +233,94 @@ public void testInvalidBBox() {
+    }
+
+    @Test
+    public void testEdgeMode() {
```
Again, what is this test supposed to test? What is 'EdgeMode'?

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -157,8 +233,94 @@ public void testInvalidBBox() {
+    }
+
+    @Test
+    public void testNodeMode() {
```
Naming again. What is 'NodeMode'?

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -157,8 +233,94 @@ public void testInvalidBBox() {
+    }
+
+    @Test
+    public void testWrongMode() {
```
Naming

sfendrich commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/centrality/ParamsTest.java:
```diff
@@ -17,30 +17,59 @@
 public class ParamsTest extends ServiceTest {
 
```
Any reason why org.json is used here, while fasterxml.jackson is used in other files (e.g., openrouteservice/src/main/java/org/heigit/ors/api/requests/centrality/CentralityRequestEnums.java)?

koebi replied:
fasterxml.jackson isn't a dependency of the openrouteservice-api-tests if I'm not mistaken.

sfendrich replied:
O.K., then we'll leave it as is for know and have to address this when cleaning up the tests.

sfendrich commented on openrouteservice/src/main/java/org/heigit/ors/api/requests/centrality/CentralityRequestHandler.java:
```diff
@@ -47,6 +47,8 @@ public CentralityResult generateCentralityFromRequest(CentralityRequest request)
 
         coreRequest.setBoundingBox(convertBBox(request.getBoundingBox()));
```
Unfortunately I cannot comment at the declaration point of this variable. Why is it called **core**Request?

koebi replied:
I think this was taken from the `MatrixRequestHandler` - compare
https://github.com/GIScience/openrouteservice/blob/5adac66e5a8a10a194406b4b18868cc426ee15e4/openrouteservice/src/main/java/org/heigit/ors/api/requests/matrix/MatrixRequestHandler.java#L41-L42

and
https://github.com/GIScience/openrouteservice/blob/5adac66e5a8a10a194406b4b18868cc426ee15e4/openrouteservice/src/main/java/org/heigit/ors/api/requests/matrix/MatrixRequestHandler.java#L53-L54

The naming in the respective functions in the `CentralityRequestHandler` is inconsistent, though:
https://github.com/GIScience/openrouteservice/blob/5adac66e5a8a10a194406b4b18868cc426ee15e4/openrouteservice/src/main/java/org/heigit/ors/api/requests/centrality/CentralityRequestHandler.java#L21-L22
vs
https://github.com/GIScience/openrouteservice/blob/5adac66e5a8a10a194406b4b18868cc426ee15e4/openrouteservice/src/main/java/org/heigit/ors/api/requests/centrality/CentralityRequestHandler.java#L33-L34

What would your suggestion be?

sfendrich replied:
I prefer the variant `centralityRequest` because for matrix 'core' also does not seem to make sense. Maybe @HendrikLeuschner knows why it is called `core`?

HendrikLeuschner replied:
My guess would be that it is the request from the API to the 'core' of the engine. I did not write these classes, though, so no idea. Agree with Sascha that `core` naming is not useful. `centralityRequest` seems better as that is also the class name. Or just go with `request`.

koebi replied:
`request` would clash with the api-request that is converted/used. I'll go with `centralityRequest` :)

sfendrich commented on openrouteservice/src/main/java/org/heigit/ors/api/responses/centrality/json/JSONCentralityLocation.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.Map;
+
+public class JSONCentralityLocation {
```
Prefer using proper CamelCase for readability: JsonCentralityLocation. Otherwise, the word boundary between JSON and Centrality is blurred. The same should be applied to many names in this PR and will not be commented seperately in this review.

koebi replied:
I agree with the word boundary blurring causing the class name to be hard to read.

`JSONCentralityLocation` was chosen to be consistent with all other response-classes, i.e. `GPXLink`, `JSONSegment` or `JSONBasedIndividualMatrixResponse`.
Any changes here wouldn't affect the rest of the code base, but would be inconsistent. Will this inconsistency be addressed by refactoring all other class names at some point?

sfendrich replied:
At least it should be addressed in the future.

koebi committed:
return warning for empty bbox

Providing an empty bounding box will yield an empty locations array.
To avoid confusion, inform the user about the empty bounding box.

koebi committed:
fix various SonarLint complaints

koebi committed:
rename centrality tests and test parameters

koebi committed:
remove todo for testing invalid excludenodes

Informing the user about a invalid excludeNode would require a
significant overhead in the calculation of nodes in the bounding box.
It could not inform the user about removing nodes he doesn't want to
remove. This has to be handled correctly by the user and is not handled
by the centrality endpoint.

koebi committed:
rename coreRequest to centralityRequest

koebi committed:
improve readability of JSON…-class names

takb committed:
Merge branch 'master' into centrality

takb committed:
Merge branch 'master' into centrality

HendrikLeuschner commented:
Earliest time I could check this is mid next week @takb

takb committed:
Merge branch 'master' into centrality
